build(docs-infra): add boilerplate to support i18n runtime example

alexcotelin opened the pull request:
This PR adds the i18n runtime example app.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

petebacondarwin commented:
@alexcotelin - thanks for continuing to progress on this. BTW the last commit should just have been an amendment or fixup of the second commit.

gkalpak commented on aio/content/examples/i18n-runtime/e2e/src/app.e2e-spec.ts:
```diff
@@ -0,0 +1,47 @@
+  });
+
+  it('should display the node texts without elements', async () => {
+    expect(await element(by.css('app-root')).getText()).toContain(
+      `Je n'affiche aucun élément`
+    );
```
```suggestion
    expect(await element(by.css('app-root')).getText()).toContain(`Je n'affiche aucun élément`);
```

gkalpak commented on aio/content/examples/i18n-runtime/e2e/src/app.e2e-spec.ts:
```diff
@@ -0,0 +1,47 @@
+  });
+
+  it('should display the ICU plural expression', async () => {
+    expect(await element.all(by.css('span')).get(0).getText()).toBe(
+      `Mis à jour à l'instant`
+    );
```
```suggestion
    expect(await element.all(by.css('span')).get(0).getText()).toBe(`Mis à jour à l'instant`);
```

gkalpak commented on aio/content/examples/i18n-runtime/e2e/src/app.e2e-spec.ts:
```diff
@@ -0,0 +1,47 @@
+    await incBtn.click();
+    await incBtn.click();
+    await element.all(by.css('button')).get(4).click();
+    expect(await nestedExp.getText()).toBe(
+      `Mis à jour: il y a 3 minutes par autre`
+    );
```
```suggestion
    expect(await nestedExp.getText()).toBe(`Mis à jour: il y a 3 minutes par autre`);
```

gkalpak commented on aio/content/examples/i18n-runtime/src/locale/messages.fr.json:
```diff
@@ -0,0 +1,14 @@
+{
```
I don't know much about i18n, but do we really need the `messages.fr.json` files in both `src/assets/` and `src/locale/`?

alexcotelin replied:
The one in `assets` it is used to load the translation when the app runs, the one in `locale` is to have it consistent to the .xlf files. @petebacondarwin do you think we should remove the one in `locale`?

petebacondarwin replied:
I think we should only have one file. One option might be to add the `locale` folder to the `asset` config in `angular.json`, given that we are modfying that file anyway?

E.g. https://github.com/angular/angular/pull/41785/files#diff-a0bdb43b061c9966fe6ac82ee46e5b25657ced3910422f6a0ff8fd9a946cf7d9R26

alexcotelin replied:
Removed it from `/assets` and loading it from `/locale` now

gkalpak commented on aio/content/examples/i18n-runtime/src/main.ts:
```diff
@@ -0,0 +1,30 @@
+}
+
+fetch('/assets/messages.fr.json')
+  .then((response) => {
```
Nit:

```suggestion
  .then(response => {
```

gkalpak commented on aio/content/examples/i18n-runtime/src/main.ts:
```diff
@@ -0,0 +1,30 @@
+fetch('/assets/messages.fr.json')
+  .then((response) => {
+    if (!response.ok) {
+      throw new Error('HTTP error ' + response.status);
```
Nit:

```suggestion
      throw new Error(`HTTP error ${response.status}`);
```

gkalpak commented on aio/content/examples/i18n-runtime/src/main.ts:
```diff
@@ -0,0 +1,30 @@
+    }
+    return response.json();
+  })
+  .then((messages) => {
```
Nit:

```suggestion
  .then(messages => {
```

gkalpak commented on aio/content/examples/i18n-runtime/src/main.ts:
```diff
@@ -0,0 +1,30 @@
+    // Bootstrap app
+    platformBrowserDynamic().bootstrapModule(AppModule);
+  })
+  .catch(() => {
+    // Error
+  });
```
Let's keep in sync with what the CLI generates by default:

```suggestion
  .catch(err => console.error(err));
```

gkalpak commented on aio/content/examples/i18n-runtime/src/app/app.module.ts:
```diff
@@ -0,0 +1,17 @@
+registerLocaleData(localeFr);
+
+@NgModule({
+  imports: [BrowserModule],
+  declarations: [AppComponent],
+  bootstrap: [AppComponent],
```
To be more consistent with other examples:

```suggestion
  imports: [ BrowserModule ],
  declarations: [ AppComponent ],
  bootstrap: [ AppComponent ],
```

petebacondarwin replied:
I think you can get this formatting in VS Code by using the "Format with... TypeScript and JavaScript Language Features" formatter.

gkalpak commented on aio/content/examples/i18n-runtime/src/app/app.component.html:
```diff
@@ -0,0 +1,37 @@
+<h1
+  i18n="User welcome|An introduction header for this sample@@introductionHeader"
+>
```
```suggestion
<h1 i18n="User welcome|An introduction header for this sample@@introductionHeader">
```

gkalpak commented on aio/content/examples/i18n-runtime/src/app/app.component.html:
```diff
@@ -0,0 +1,37 @@
+<br />
+<button (click)="inc(1)">+</button> <button (click)="inc(-1)">-</button>
+
+<span i18n
+  >Updated {minutes, plural, =0 {just now} =1 {one minute ago} other
+  {{{minutes}} minutes ago}}</span
+>
```
FWIW, I find this way of formatting difficult to read. I'd rather stick to how similar elements are formatted in the `i18n` examples (but I don't feel particularly strongly about it).

gkalpak commented on aio/content/examples/i18n-runtime/example-config.json:
```diff
@@ -0,0 +1,14 @@
+{
+  "projectType": "i18n-runtime",
+  "tests": [
+    {
+      "cmd": "yarn",
+      "args": [
+        "e2e",
+        "--protractor-config=e2e/protractor-puppeteer.conf.js",
+        "--no-webdriver-update",
+        "--port={PORT}"
+      ]
+    }
+  ]
```
Is the `tests` property necessary here? (The only difference from the default test command is that the default also passes the `--prod` option. Does the example fail with the `--prod` option?)

gkalpak reviewed (changes requested):
Thx for working on this, @alexcotelin :+1:

I have left some (mostly minor) comments from a `docs-infra` perspective. Can you please also add the `i18n-runtime` example to the `fw-i18n` group in `.pullapprove.yml` ([similar to how we do for the `i18n` one][1])?

[1]: https://github.com/angular/angular/blob/33c7f28baf15925bc53a64b9fa89920ebf09dfd1/.pullapprove.yml#L564

gkalpak commented on aio/content/examples/i18n-runtime/example-config.json:
```diff
@@ -1,14 +1 @@
-    }
-  ]
-}
+
```
Don't we need to set `projectType` to `i18n-runtime`? 😕

petebacondarwin replied:
I hope that the example's tests would fail without this! It would be a good check of whether the tests are meaningful or not. I'll try running CI for this off head...

alexcotelin replied:
Added `projectType` back.

gkalpak reviewed (changes requested):
Thx for making the changes, @alexcotelin :heart:
Just a couple of outstanding comments and this should be good to go 🚀 

##
<sub>

[NOTE TO SELF] Unresolved comments from previous review: https://github.com/angular/angular/pull/41785#discussion_r627456700, https://github.com/angular/angular/pull/41785#discussion_r627468372

</sub>

petebacondarwin commented on aio/content/examples/i18n-runtime/stackblitz.json:
```diff
@@ -0,0 +1,12 @@
+    "!**/*.js",
+    "!**/*.[0-9].*",
+    "!doc-files/**/*",
+    "**/*.xlf"
```
I think we need to include the XLIFF files in the StackBlitz, no?

petebacondarwin replied:
Is this needed? Won't it include these by default any way, since they are not included?

petebacondarwin commented on aio/tools/examples/shared/boilerplate/i18n-runtime/angular.json:
```diff
@@ -0,0 +1,116 @@
+{
```
I think it would be great to add some comments to this file to indicate how it has been changed from the default. This would also be helpful when updating the Angular versions of the examples.

alexcotelin replied:
Same as cli just added `"src/locale"` line 28

mary-poppins commented:
You can preview 6a17fab at https://pr41785-6a17fab.ngbuilds.io/.
You can preview 1e27945 at https://pr41785-1e27945.ngbuilds.io/.

alexcotelin committed:
docs: add i18n-runtime example app

alexcotelin committed:
fixup! docs: add i18n-runtime example app

alexcotelin committed:
build(docs-infra): add boilerplate to support i18n runtime example

Existing boilerplate do not work for an i18n runtime example

gkalpak commented on aio/content/examples/i18n-runtime/stackblitz.json:
```diff
@@ -0,0 +1,12 @@
+  ],
+  "file": "src/app/app.component.ts",
+  "tags": ["Angular", "i18n", "internationalization", "runtime"]
+}
```
Do we need to add `@angular/compiler-cli` as a devDependency (as introduced in #42001 - see https://github.com/angular/angular/pull/42001/files#diff-4e76265e2bd293666e42fd9f2202b87693ef5351daa9e97e78a53bfc586a9c4eR12)?

petebacondarwin replied:
We will need such a config, now that #42001 landed.

gkalpak reviewed (changes requested):
One last comment (in addition to @petebacondarwin's). Otherwise lgtm ✨
(FYI, I have rebased this on latest master to fix CI.)

mary-poppins commented:
You can preview 7b32513 at https://pr41785-7b32513.ngbuilds.io/.

alexcotelin committed:
fixup! docs: add i18n-runtime example app

mary-poppins commented:
You can preview 4f0331e at https://pr41785-4f0331e.ngbuilds.io/.

alxhub reviewed (approved):
Reviewed-for: code-ownership

mary-poppins commented:
You can preview 5356b92 at https://pr41785-5356b92.ngbuilds.io/.

mary-poppins commented:
You can preview 1536a31 at https://pr41785-1536a31.ngbuilds.io/.

petebacondarwin commented:
@alexcotelin - I have rebased this on master and there are now legitimate failures in the new e2e tests. Are you able to look into it or do you need help?

CDDelta commented:
Hi, I'm trying to implement runtime i18n with this PR and I'm curious as to how `messages.fr.json` is generated from `messages.fr.xlf` (at least that is what I am assuming is being done). Could I get some pointers here? Thank you.

alexcotelin commented:
> @alexcotelin - I have rebased this on master and there are now legitimate failures in the new e2e tests. Are you able to look into it or do you need help?

Hi yes I'll have a look this weekend

alexcotelin commented:
> Hi, I'm trying to implement runtime i18n with this PR and I'm curious as to how `messages.fr.json` is generated from `messages.fr.xlf` (at least that is what I am assuming is being done). Could I get some pointers here? Thank you.

You can pass `--format=json` to extract-i18n command [https://angular.io/cli/extract-i18n](https://angular.io/cli/extract-i18n )

CDDelta commented:
> You can pass `--format=json` to extract-i18n command https://angular.io/cli/extract-i18n

This extracts the content from the app though right? I am looking for a way to convert an existing `.xlf` file to a `.json` file that can be consumed by Angular :)

petebacondarwin commented:
> > You can pass `--format=json` to extract-i18n command https://angular.io/cli/extract-i18n
> 
> This extracts the content from the app though right? I am looking for a way to convert an existing `.xlf` file to a `.json` file that can be consumed by Angular :)

If you are will to do a bit of coding (and accessing our private APIs) it is quite straightforward... something like this: https://stackblitz.com/edit/node-gvmexg?file=index.js

CDDelta commented:
> > > You can pass `--format=json` to extract-i18n command https://angular.io/cli/extract-i18n
> > 
> > 
> > This extracts the content from the app though right? I am looking for a way to convert an existing `.xlf` file to a `.json` file that can be consumed by Angular :)
> 
> If you are will to do a bit of coding (and accessing our private APIs) it is quite straightforward... something like this: https://stackblitz.com/edit/node-gvmexg?file=index.js

Thanks @petebacondarwin for the example, it is very appreciated! Would it make sense to expose this in the CLI? I can see this being very useful for those using the default `.xlf` files but wanting to do runtime i18n as well. I can file the request if it makes sense to.

alexcotelin commented:
> @alexcotelin - I have rebased this on master and there are now legitimate failures in the new e2e tests. Are you able to look into it or do you need help?

So had a look and the problem is that we removed ` aio/tools/examples/shared/boilerplate/i18n-runtime/angular.json` and now the app fails to start because the `locale/messages.fr.json` is missing, so we need to include `src/locale` in `angular.json` build.options.assets. Do I add it `angular.json` back? also before we had the `messages.fr.json` in the `assets` folder

petebacondarwin commented:
I think you probably should add the angular.json file back, unless it is identical to the on in the cli boilerplate.

mary-poppins commented:
You can preview 05c5d94 at https://pr41785-05c5d94.ngbuilds.io/.

AndrewKushnir commented:
Closing this PR due to the lack of response. Please reopen when you get a chance to look into this. Thank you.
